File: Day4/part2/main.py
def read_file():
    """Reads the input file and returns a 2D list of characters."""
    with open("input.txt", "r") as file:
        return [list(line.strip()) for line in file]

def check_bounds(rows, cols, x, y):
    """Checks if the coordinates (x, y) are within the grid boundaries."""
    return 0 <= x < rows and 0 <= y < cols

def check_xmas_pattern(top_left, top_right, bottom_left, bottom_right):
    """Checks if the given characters form a valid XMAS pattern."""
    return ((top_left == 'M' and top_right == 'M' and bottom_left == 'S' and bottom_right == 'S') or
            (top_left == 'S' and top_right == 'S' and bottom_left == 'M' and bottom_right == 'M') or
            (top_left == 'M' and top_right == 'S' and bottom_left == 'M' and bottom_right == 'S') or
            (top_left == 'S' and top_right == 'M' and bottom_left == 'S' and bottom_right == 'M'))

def search_xmas(grid):
    """Searches for custom XMAS patterns around the character 'A'."""
    rows = len(grid)
    cols = len(grid[0])
    matches = 0

    for x in range(rows):
        for y in range(cols):
            if grid[x][y] != 'A':
                continue

            top_left_x, top_left_y = x - 1, y - 1
            top_right_x, top_right_y = x - 1, y + 1
            bottom_left_x, bottom_left_y = x + 1, y - 1
            bottom_right_x, bottom_right_y = x + 1, y + 1

            if not (check_bounds(rows, cols, top_left_x, top_left_y) and
                    check_bounds(rows, cols, top_right_x, top_right_y) and
                    check_bounds(rows, cols, bottom_left_x, bottom_left_y) and
                    check_bounds(rows, cols, bottom_right_x, bottom_right_y)):
                continue

            top_left = grid[top_left_x][top_left_y]
            top_right = grid[top_right_x][top_right_y]
            bottom_left = grid[bottom_left_x][bottom_left_y]
            bottom_right = grid[bottom_right_x][bottom_right_y]

            if check_xmas_pattern(top_left, top_right, bottom_left, bottom_right):
                matches += 1

    return matches

def part2():
    grid = read_file()
    matches = search_xmas(grid)
    print("Part 2 answer:", matches)

if __name__ == "__main__":
    part2()

File: Day4/part2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// Read input from file and convert it to a 2D rune slice
func readFile() [][]rune {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var input [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		input = append(input, []rune(text))
	}

	return input
}

// Check if the surrounding characters form a valid "XMAS" pattern
func checkXmasPattern(topLeft, topRight, bottomLeft, bottomRight rune) bool {
	return (topLeft == 'M' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'S') ||
		(topLeft == 'S' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'M') ||
		(topLeft == 'M' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'S') ||
		(topLeft == 'S' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'M')
}

// Check if coordinates are within bounds
func checkBounds(rows, cols, x, y int) bool {
	return x >= 0 && x < rows && y >= 0 && y < cols
}

// Search for custom "XMAS" patterns
func searchXmas(grid [][]rune) int {
	rows := len(grid)
	cols := len(grid[0])
	matches := 0

	for x, row := range grid {
		for y := range row {
			if grid[x][y] != 'A' {
				continue
			}

			topLeftX, topLeftY := x-1, y-1
			topRightX, topRightY := x-1, y+1
			bottomLeftX, bottomLeftY := x+1, y-1
			bottomRightX, bottomRightY := x+1, y+1

			if !checkBounds(rows, cols, topLeftX, topLeftY) ||
				!checkBounds(rows, cols, topRightX, topRightY) ||
				!checkBounds(rows, cols, bottomLeftX, bottomLeftY) ||
				!checkBounds(rows, cols, bottomRightX, bottomRightY) {
				continue
			}

			topLeft := grid[topLeftX][topLeftY]
			topRight := grid[topRightX][topRightY]
			bottomLeft := grid[bottomLeftX][bottomLeftY]
			bottomRight := grid[bottomRightX][bottomRightY]

			if checkXmasPattern(topLeft, topRight, bottomLeft, bottomRight) {
				matches++
			}
		}
	}

	return matches
}

// Part 2: Search for custom "XMAS" patterns
func part2() {
	grid := readFile()
	matches := searchXmas(grid)
	fmt.Println("Part 2 answer:", matches)
}

func main() {
	part2()
}
